mars-sim-core/src/main/java/com/mars_sim/core/structure/building/BuildingManager.java

```diff
@@ -2031,33 +1969,39 @@ public double getTotalEntropyPerLab() {
 	public Research getWorstEntropyLabByProbability(Person person) {
 		Map<Research, Double> scores = new HashMap<>();
 		Set<Building> bldgs = getBuildingSet(FunctionType.RESEARCH);
+				
+		if (bldgs.isEmpty())
+			return null;
+
+		for (Building b: bldgs) {
+			Research lab = b.getResearch();
+			double entropy = lab.getEntropy();
+			scores.put(lab, entropy);
+		}
+		return getWorstEntropyByProbability(person, bldgs, scores);
+	}
+
+	public <T> T getWorstEntropyByProbability(Person person, Set<Building> buildings,
+											   Map<T, Double> scores) {
+
 		if (person.getBuildingLocation() != null) {
-			bldgs = bldgs
+			buildings = buildings
 					.stream()
 					.filter(b -> b.getZone() == person.getBuildingLocation().getZone()
 							&& !b.getMalfunctionManager().hasMalfunction())
 					.collect(Collectors.toSet());
 		}
 		else {
-			bldgs = bldgs
+			buildings = buildings
 					.stream()
 					.filter(b -> b.getZone() == 0
```

Comment:
As a sidenote, we haven't really applied the concept of district division or zoning much for building management yet.

But we'll revisit the need of zoning as the settlement grows to over a hundred buildings.